File: mobile/lib/models/complaint.dart
import 'category.dart';

class ComplaintListItem {
  final String id;
  final String complaintNumber;
  final String categoryCode;
  final String description;
  final String severity;
  final String status;
  final double latitude;
  final double longitude;
  final DateTime createdAt;
  final DateTime? deadlineAt;

  ComplaintListItem({
    required this.id,
    required this.complaintNumber,
    required this.categoryCode,
    required this.description,
    required this.severity,
    required this.status,
    required this.latitude,
    required this.longitude,
    required this.createdAt,
    required this.deadlineAt,
  });

  factory ComplaintListItem.fromJson(Map<String, dynamic> json) {
    return ComplaintListItem(
      id: json['id'] as String,
      complaintNumber: json['complaint_number'] as String,
      categoryCode: json['category_code'] as String,
      description: json['description'] as String,
      severity: json['severity'] as String,
      status: json['status'] as String,
      latitude: (json['latitude'] as num).toDouble(),
      longitude: (json['longitude'] as num).toDouble(),
      createdAt: DateTime.parse(json['created_at'] as String),
      deadlineAt: json['deadline_at'] != null ? DateTime.parse(json['deadline_at'] as String) : null,
    );
  }
}

class ComplaintDetail {
  final String id;
  final String complaintNumber;
  final ComplaintCategory category;
  final String description;
  final String? imageUrl;
  final double latitude;
  final double longitude;
  final String? addressText;
  final String severity;
  final String status;
  final String? aiCategoryCode;
  final double? aiConfidence;
  final DateTime? deadlineAt;
  final int escalationLevel;
  final String? rejectionReason;
  final DateTime createdAt;

  ComplaintDetail({
    required this.id,
    required this.complaintNumber,
    required this.category,
    required this.description,
    required this.imageUrl,
    required this.latitude,
    required this.longitude,
    required this.addressText,
    required this.severity,
    required this.status,
    required this.aiCategoryCode,
    required this.aiConfidence,
    required this.deadlineAt,
    required this.escalationLevel,
    required this.rejectionReason,
    required this.createdAt,
  });

  factory ComplaintDetail.fromJson(Map<String, dynamic> json) {
    return ComplaintDetail(
      id: json['id'] as String,
      complaintNumber: json['complaint_number'] as String,
      category: ComplaintCategory.fromJson(json['category'] as Map<String, dynamic>),
      description: json['description'] as String,
      imageUrl: json['image_url'] as String?,
      latitude: (json['latitude'] as num).toDouble(),
      longitude: (json['longitude'] as num).toDouble(),
      addressText: json['address_text'] as String?,
      severity: json['severity'] as String,
      status: json['status'] as String,
      aiCategoryCode: json['ai_category_code'] as String?,
      aiConfidence: (json['ai_confidence'] as num?)?.toDouble(),
      deadlineAt: json['deadline_at'] != null ? DateTime.parse(json['deadline_at'] as String) : null,
      escalationLevel: json['escalation_level'] as int,
      rejectionReason: json['rejection_reason'] as String?,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }
}

File: backend/app/schemas/complaint.py
import uuid
from datetime import datetime

from pydantic import BaseModel, Field


class ComplaintCategoryOut(BaseModel):
    id: uuid.UUID
    code: str
    name: str
    module_id: str
    default_severity: str
    icon: str

    model_config = {"from_attributes": True}


class ComplaintCreate(BaseModel):
    category_code: str
    description: str = Field(min_length=5, max_length=3000)
    latitude: float = Field(ge=-90, le=90)
    longitude: float = Field(ge=-180, le=180)
    address_text: str | None = None
    extra_data: dict | None = None


class DuplicateHint(BaseModel):
    complaint_id: uuid.UUID
    complaint_number: str
    score: float


class ComplaintOut(BaseModel):
    id: uuid.UUID
    complaint_number: str
    user_id: uuid.UUID
    category: ComplaintCategoryOut
    description: str
    image_url: str | None
    latitude: float
    longitude: float
    address_text: str | None
    severity: str
    ai_category_code: str | None
    ai_confidence: float | None
    ai_severity: str | None
    status: str
    assigned_authority_id: uuid.UUID | None
    deadline_at: datetime | None
    escalation_level: int
    duplicate_of_id: uuid.UUID | None
    rejection_reason: str | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ComplaintListItem(BaseModel):
    id: uuid.UUID
    complaint_number: str
    category_code: str
    description: str
    severity: str
    status: str
    latitude: float
    longitude: float
    created_at: datetime
    deadline_at: datetime | None


class StatusUpdateRequest(BaseModel):
    to_status: str
    note: str | None = None


class ResolutionCreate(BaseModel):
    description: str = Field(min_length=5, max_length=2000)
    latitude: float | None = None
    longitude: float | None = None


class ResolutionOut(BaseModel):
    id: uuid.UUID
    complaint_id: uuid.UUID
    description: str
    photo_url: str | None
    resolved_by: uuid.UUID
    resolved_at: datetime
    citizen_verified: bool | None
    citizen_feedback: str | None

    model_config = {"from_attributes": True}


class CitizenVerificationRequest(BaseModel):
    verified: bool
    feedback: str | None = None


class RejectionRequest(BaseModel):
    reason: str = Field(min_length=5, max_length=1000)
